File: webhooks/webhook_reader.py
import cgi
import json
import os
import urllib.parse
from http.server import BaseHTTPRequestHandler, HTTPServer
from json.decoder import JSONDecodeError

upload_to = "uploads"


class GetHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b"Send a POST request instead.")
        return

    def do_POST(self):
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        ctype, pdict = cgi.parse_header(self.headers["Content-Type"])
        if ctype == "multipart/form-data":
            form = cgi.FieldStorage(
                fp=self.rfile,
                headers=self.headers,
                environ={
                    "REQUEST_METHOD": "POST",
                    "CONTENT_TYPE": self.headers["Content-Type"],
                },
            )
            # Get webhook content
            json_data = form.getvalue("json")
            # Get webhook file
            if "upload" in form:
                filename = form["upload"].filename
                buffer = form["upload"].file.read()
                if not os.path.exists(upload_to):
                    try:
                        os.makedirs(upload_to)
                    except OSError:
                        print("Error creating directory:", upload_to)

                with open(f"./{upload_to}/{filename}", "wb") as fp:
                    print("Saving image to %s" % filename)
                    fp.write(buffer)

        else:
            raw_data = self.rfile.read(int(self.headers["content-length"])).decode(
                "utf-8"
            )
            if raw_data.startswith("json="):
                raw_data = raw_data[5:]
            try:
                decoded_data = urllib.parse.unquote(raw_data)
                decoded_data = decoded_data.replace("+", " ")
                json_data = json.loads(decoded_data)
            except JSONDecodeError:
                json_data = {}
        print(json_data)
        self.wfile.write(b"OK")


if __name__ == "__main__":
    server = HTTPServer(("", 8001), GetHandler)
    print("Starting server, use <Ctrl-C> to stop")
    server.serve_forever()

File: webhooks/webhook_reader.js
/**
 * # Install dependecies
 *
 * npm install
 *
 * # Run
 * node webhook_reader.js
 */

const fs = require("node:fs");
const http = require("node:http");
const multer = require("multer");

const uploadsDir = "./uploads";

const storage = multer.diskStorage({
  destination: (_req, _file, cb) => {
    if (!fs.existsSync(uploadsDir)) {
      fs.mkdirSync(uploadsDir);
    }
    cb(null, uploadsDir);
  },
  filename: (_req, file, cb) => {
    cb(null, file.originalname);
  },
});

const filesParser = multer({ storage }).any();

const server = http.createServer((req, res) => {
  if (req.method === "POST") {
    collectRequestData(req, res);
  } else {
    res.end(`Send a POST request instead.`);
  }
});

server.listen(8001, () => {
  console.log("Server listening on port 8001");
});

const CONTENT_TYPE_MULTIPART_FORM_DATA = "multipart/form-data";

function collectRequestData(request, response) {
  const contentType = request.headers["content-type"];
  // console.log(`Content type: ${contentType}`);
  if (!contentType) {
    response.end("OK!");
  } else if (contentType.indexOf(CONTENT_TYPE_MULTIPART_FORM_DATA) > -1) {
    filesParser(request, response, (err) => {
      if (err) {
        // A Multer error occurred when uploading.
        // An unknown error occurred when uploading.
        console.error(err);
        response.writeHead(500);
        response.end("Error");
      } else {
        jsonData = request.body.json;
        console.log(jsonData);
        response.end("OK!");
      }
    });
  } else {
    let rawData = "";
    request.on("data", (data) => {
      rawData += data;
    });
    request.on("end", () => {
      let decodedData = decodeURIComponent(rawData);
      if (decodedData.includes("json=")) {
        decodedData = decodedData.split("+").join(" ");
        const jsonData = decodedData.split("json=")[1];
        console.log(jsonData);
        response.end("OK!");
      }
    });
  }
}
